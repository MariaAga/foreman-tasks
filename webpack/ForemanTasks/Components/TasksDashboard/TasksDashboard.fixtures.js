import { selectTasksSummary } from './TasksDashboardSelectors';

export const tasksSummary = {
  running: {
    recent: 1,
    total: 3,
  },
  paused: {
    recent: 3,
    total: 4,
  },
  stopped: {
    by_result: {
      error: {
        total: 5,
        recent: 6,
      },
      warning: {
        total: 7,
        recent: 8,
      },
      success: {
        total: 9,
        recent: 10,
      },
    },
  },
  scheduled: {
    total: 11,
  },
};

export const tasksDashboardProps = {
  tasksSummary: selectTasksSummary({
    foremanTasks: {
      tasksDashboard: { tasksSummary },
    },
  }),
};
