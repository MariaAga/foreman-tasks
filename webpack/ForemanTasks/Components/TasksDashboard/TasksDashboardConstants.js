import { translate as __ } from 'foremanReact/common/I18n';

export const TASKS_SUMMARY_REQUEST = 'TASKS_SUMMARY_REQUEST';
export const TASKS_SUMMARY_SUCCESS = 'TASKS_SUMMARY_SUCCESS';
export const TASKS_SUMMARY_FAILURE = 'TASKS_SUMMARY_FAILURE';

export const TASKS_SUMMARY_ZERO = {
  running: {
    recent: 0,
    total: 0,
  },
  paused: {
    recent: 0,
    total: 0,
  },
  stopped: {
    by_result: {
      error: {
        total: 0,
        recent: 0,
      },
      warning: {
        total: 0,
        recent: 0,
      },
      success: {
        total: 0,
        recent: 0,
      },
    },
  },
  scheduled: {
    total: 0,
  },
};

export const TASKS_DASHBOARD_AVAILABLE_QUERY_STATES = {
  RUNNING: 'running',
  PAUSED: 'paused',
  STOPPED: 'stopped',
  SCHEDULED: 'scheduled',
};

export const TASKS_DASHBOARD_AVAILABLE_QUERY_RESULTS = {
  ERROR: 'error',
  WARNING: 'warning',
  SUCCESS: 'success',
};

export const TASKS_DASHBOARD_AVAILABLE_QUERY_MODES = {
  LAST: 'last',
  OLDER: 'older',
};

export const TASKS_DASHBOARD_AVAILABLE_TIMES = {
  H24: 'H24',
  H12: 'H12',
  WEEK: 'WEEK',
};

export const TASKS_DASHBOARD_AVAILABLE_TIMES_TEXT = {
  [TASKS_DASHBOARD_AVAILABLE_TIMES.H24]: __('24h'),
  [TASKS_DASHBOARD_AVAILABLE_TIMES.H12]: __('12h'),
  [TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK]: __('week'),
};

export const STATUS = {
  PENDING: 'PENDING',
  RESOLVED: 'RESOLVED',
  ERROR: 'ERROR',
};
