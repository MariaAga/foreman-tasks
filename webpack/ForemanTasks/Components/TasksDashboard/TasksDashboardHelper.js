import URI from 'urijs';
import { getURIQuery } from 'foremanReact/common/helpers';

import {
  TASKS_DASHBOARD_AVAILABLE_TIMES,
  TASKS_DASHBOARD_QUERY_KEYS_TEXT,
  TASKS_DASHBOARD_QUERY_VALUES_TEXT,
} from './TasksDashboardConstants';

export const getQueryKeyText = key => TASKS_DASHBOARD_QUERY_KEYS_TEXT[key];

export const getQueryValueText = value =>
  TASKS_DASHBOARD_QUERY_VALUES_TEXT[value];

export const timeToHoursNumber = time => {
  switch (time) {
    case TASKS_DASHBOARD_AVAILABLE_TIMES.H12:
      return 12;
    case TASKS_DASHBOARD_AVAILABLE_TIMES.H24:
      return 24;
    case TASKS_DASHBOARD_AVAILABLE_TIMES.WEEK:
      return 24 * 7;
    default:
      return 24;
  }
};

const uriToQueryMap = {
  state: 'state',
  result: 'result',
  time_mode: 'mode',
  time_horizon: 'time',
};

const queryFromUriQuery = uriQuery => {
  const query = {};

  Object.entries(uriToQueryMap).forEach(([uriField, queryField]) => {
    if (uriQuery[uriField]) query[queryField] = uriQuery[uriField];
  });

  if (query.mode === 'recent') {
    query.mode = 'last';
  }

  return query;
};

const uriQueryFromQuery = query => {
  const uriQuery = {};

  Object.entries(uriToQueryMap).forEach(([uriField, queryField]) => {
    if (query[queryField]) uriQuery[uriField] = query[queryField];
  });

  if (uriQuery.time_mode === 'last') {
    uriQuery.time_mode = 'recent';
  }

  return uriQuery;
};

export const getQueryFromUrl = () => {
  const uriQuery = getURIQuery(window.location.href);

  return queryFromUriQuery(uriQuery);
};

export const updateURLQuery = ({ state, result, mode, time, uri }) => {
  const query = {
    state,
    result,
    time_mode: mode === 'last' ? 'recent' : mode,
    time_horizon: time,
    page: 1,
  };

  uri.setSearch(query);
  window.history.pushState({ path: uri.toString() }, '', uri.toString());
};

export const resolveQuery = query => {
  const uriQuery = uriQueryFromQuery(query);

  const uri = new URI(window.location.href);
  const { search, per_page: perPage } = uri.query(true);

  const data = { search, ...uriQuery, page: 1, per_page: perPage };
  const { $, tfm } = window;

  updateURLQuery({ uri, ...query });

  uri.query(URI.buildQuery(data, true));
  tfm.tools.showSpinner();
  $.ajax({
    type: 'get',
    url: uri.toString(),
    success(result) {
      const res = $(`<div>${result}</div>`);

      $('#tasks-table').html(res.find('#tasks-table'));
    },
    error({ statusText }) {
      $('#tasks-table').html(statusText);
    },
    complete(result) {
      tfm.tools.hideSpinner();
    },
  });
};
