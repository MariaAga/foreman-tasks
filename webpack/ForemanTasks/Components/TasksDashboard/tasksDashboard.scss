.tasks-dashboard {
  display: grid;
  grid-template-columns: 25% 25% 25% 25%;

  .card-pf {
    width: 300px;
  }
  // .tasks-donut-card {
  //   display: flex;
  //   justify-content: center;
  // }
  // .column {
  //   display: flex;
  //   flex-direction: column;
  //   flex-basis: 100%;
  //   flex: 1;
  // }
}
