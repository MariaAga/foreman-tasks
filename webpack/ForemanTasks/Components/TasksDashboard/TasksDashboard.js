import React from 'react';
import PropTypes from 'prop-types';
import RunningTasksCard from './Components/RunningTasksCard/RunningTasksCard';
import PausedTasksCard from './Components/PausedTasksCard/PausedTasksCard';
import StoppedTasksCard from './Components/StoppedTasksCard/StoppedTasksCard';
import ScheduledTasksCard from './Components/ScheduledTasksCard/ScheduledTasksCard';
import { buildObject } from './TasksDashboardHelper';

class TasksDashboard extends React.Component {
  state = {
    focusedOn: this.props.focusedOn,
  };
  updateQuery = query => {
    const { state, time: timePeriod, result } = query;
    const data = [state, result, timePeriod || 'total'].filter(Boolean);
    const focusedOn = buildObject(data);
    console.log(focusedOn);
    this.setState({
      focusedOn,
    });
  };
  onTotalClickRunning = () => this.updateQuery({ state: 'running' });
  onTotalClickPaused = () => this.updateQuery({ state: 'paused' });
  onLastClickRunning = () =>
    this.updateQuery({ state: 'running', time: 'last' });
  onOlderClickRunning = () =>
    this.updateQuery({ state: 'running', time: 'older' });
  onLastClickPaused = () => this.updateQuery({ state: 'paused', time: 'last' });
  onOlderClickPaused = () =>
    this.updateQuery({ state: 'paused', time: 'older' });

  getFunctions = data => {
    data.updateQuery = this.updateQuery;
    data.stopped.func = () =>
      this.updateQuery({
        state: 'stopped',
      });
    for (const result of Object.keys(data.stopped)) {
      for (const time of Object.keys(data.stopped[result])) {
        data.stopped[result][time].func = () =>
          this.updateQuery({
            state: 'stopped',
            result,
            time,
          });
      }
    }
    return data;
  };
  getMockData = () => ({
    timePeriod: '24h',
    running: {
      last: 3,
      older: 5,
    },
    paused: {
      last: 3,
      older: 5,
    },
    stopped: {
      error: {
        total: {
          value: 8,
        },
        last: {
          value: 1,
        },
      },
      warning: {
        total: {
          value: 20,
        },
        last: {
          value: 2,
        },
      },
      success: {
        total: {
          value: 25,
        },
        last: {
          value: 3,
        },
      },
    },
    scheduled: {
      scheduled: 1,
    },
  });

  render() {
    const data = this.getFunctions(this.getMockData());
    const { timePeriod, updateQuery } = data;
    const { focusedOn } = this.state;
    return (
      <div className="tasks-dashboard">
        <RunningTasksCard
          updateQuery={updateQuery}
          timePeriod={timePeriod}
          focusedOn={focusedOn.running}
          onTotalClick={this.onTotalClickRunning}
          onLastClick={this.onLastClickRunning}
          onOlderClick={this.onOlderClickRunning}
          {...data.running}
        />

        <PausedTasksCard
          updateQuery={updateQuery}
          timePeriod={timePeriod}
          focusedOn={focusedOn.paused}
          onTotalClick={this.onTotalClickPaused}
          onLastClick={this.onLastClickPaused}
          onOlderClick={this.onOlderClickPaused}
          {...data.paused}
        />

        <StoppedTasksCard
          timePeriod={timePeriod}
          focusedOn={focusedOn.stopped}
          {...data.stopped}
        />

        <ScheduledTasksCard
          updateQuery={updateQuery}
          timePeriod={timePeriod}
          focusedOn={focusedOn.scheduled}
          {...data.scheduled}
        />
      </div>
    );
  }
}

TasksDashboard.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  focusedOn: PropTypes.shape({
    running: {
      last: PropTypes.bool,
      older: PropTypes.bool,
      total: PropTypes.bool,
    },
    paused: {
      last: PropTypes.bool,
      older: PropTypes.bool,
      total: PropTypes.bool,
    },
    stopped: {
      error: {
        total: PropTypes.bool,
        last: PropTypes.bool,
      },
      warning: {
        total: PropTypes.bool,
        last: PropTypes.bool,
      },
      success: {
        total: PropTypes.bool,
        last: PropTypes.bool,
      },
    },
    scheduled: {
      scheduled: PropTypes.bool,
    },
  }),
};

TasksDashboard.defaultProps = {
  className: '',
  children: null,
  focusedOn: {
    running: { normal: true },
    paused: { normal: true },
    stopped: { normal: true },
    scheduled: { normal: true },
  },
};

export default TasksDashboard;
