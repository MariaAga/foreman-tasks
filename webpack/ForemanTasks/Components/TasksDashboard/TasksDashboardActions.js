import API from 'foremanReact/API';
import { resolveQuery } from './TasksDashboardHelper';
import {
  FOREMAN_TASKS_DASHBOARD_INIT,
  FOREMAN_TASKS_DASHBOARD_UPDATE_TIME,
  FOREMAN_TASKS_DASHBOARD_UPDATE_QUERY,
  TASKS_DASHBOARD_CURRENT_TIME,
  TASKS_SUMMARY_FAILURE,
  TASKS_SUMMARY_REQUEST,
  TASKS_SUMMARY_SUCCESS,
  STATUS,
} from './TasksDashboardConstants';
import { selectTime } from './TasksDashboardSelectors';

export const initializeDashboard = ({ time, query }) => ({
  type: FOREMAN_TASKS_DASHBOARD_INIT,
  payload: { time, query },
});

export const updateTime = time => ({
  type: FOREMAN_TASKS_DASHBOARD_UPDATE_TIME,
  payload: time,
});

export const updateQuery = query => (dispatch, getState) => {
  if (query.time === TASKS_DASHBOARD_CURRENT_TIME)
    query.time = selectTime(getState());

  dispatch({
    type: FOREMAN_TASKS_DASHBOARD_UPDATE_QUERY,
    payload: query,
  });

  resolveQuery(query);
};

export const getTasksSummary = () => dispatch => {
  dispatch(startRequest());
  return createAPIRequest({ dispatch });
};

const startRequest = () => ({
  type: TASKS_SUMMARY_REQUEST,
  payload: {
    status: STATUS.PENDING,
  },
});

const createAPIRequest = ({ dispatch }) =>
  API.get('/foreman_tasks/tasks/summary/24')
    .then(({ data }) =>
      dispatch(
        requestSuccess({
          data,
        })
      )
    )
    .catch(error => dispatch(requestFailure({ error })));

const requestFailure = ({ error }) => ({
  type: TASKS_SUMMARY_FAILURE,
  payload: {
    results: [],
    error: error.message || error,
    status: STATUS.ERROR,
  },
});

const requestSuccess = ({ data }) => ({
  type: TASKS_SUMMARY_SUCCESS,
  payload: {
    ...data,
  },
});
