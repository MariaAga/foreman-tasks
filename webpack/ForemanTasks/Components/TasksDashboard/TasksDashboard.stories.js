import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, object, boolean } from '@storybook/addon-knobs';

import TasksDashboard from './TasksDashboard';

import './tasksDashboard.scss';

storiesOf('TasksDashboard', module)
  .addDecorator(withKnobs)
  .add('TasksDashboard', () => (
    <div>
      <link
        href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css"
        rel="stylesheet"
        integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN"
        crossOrigin="anonymous"
      />
      <TasksDashboard
        focusedOn={object('Focused on', {
          running: object('running', {
            normal: boolean('Running normal', true),

            recent: boolean('Running recent', false),
            total: boolean('Running total', false),
          }),
          paused: object('paused', {
            normal: boolean('Paused normal', true),

            recent: boolean('Paused recent', false),
            total: boolean('Paused total', false),
          }),
          stopped: object('stopped', {
            normal: boolean('Stopped normal', true),

            success: object('success', {
              recent: boolean('recent', false),
              total: boolean('total', false),
            }),
            error: object('error', {
              recent: boolean('recent', false),
              total: boolean('total', false),
            }),
            warning: object('warning', {
              recent: boolean('recent', false),
              total: boolean('total', false),
            }),
          }),
          scheduled: object('scheduled', {
            normal: boolean('Scheduled normal', true),
            total: boolean('Scheduled total', false),
          }),
        })}
      />
    </div>
  ));
