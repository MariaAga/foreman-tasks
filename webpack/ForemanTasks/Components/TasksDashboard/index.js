import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import TasksDashboard from './TasksDashboard';
import * as actions from './TasksDashboardActions';
import reducer from './TasksDashboardReducer';
import {
  selectTimePeriod,
  selectTasksSummary,
} from './TasksDashboardSelectors';
import './TasksDashboard.scss';

const mapStateToProps = state => ({
  timePeriod: selectTimePeriod(state),
  tasksSummary: selectTasksSummary(state),
});

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { tasksDashboard: reducer };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TasksDashboard);
