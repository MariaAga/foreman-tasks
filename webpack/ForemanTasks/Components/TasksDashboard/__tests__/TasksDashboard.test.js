import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import TasksDashboard from '../TasksDashboard';

const fixtures = {
  'render without Props': { history: {} },
  /** fixtures, props for the component */
};

describe('TasksDashboard', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TasksDashboard, fixtures));
});
