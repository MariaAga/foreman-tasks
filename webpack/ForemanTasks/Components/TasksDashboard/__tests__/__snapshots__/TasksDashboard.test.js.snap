// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksDashboard rendering render without Props 1`] = `
<div
  className="tasks-dashboard"
>
  <RunningTasksCard
    className=""
    focusedOn={
      Object {
        "normal": true,
      }
    }
    last={3}
    older={5}
    onLastClick={[Function]}
    onOlderClick={[Function]}
    onTotalClick={[Function]}
    timePeriod="24h"
  />
  <PausedTasksCard
    className=""
    focusedOn={
      Object {
        "normal": true,
      }
    }
    last={3}
    older={5}
    onLastClick={[Function]}
    onOlderClick={[Function]}
    onTotalClick={[Function]}
    timePeriod="24h"
  />
  <StoppedTasksCard
    className=""
    error={
      Object {
        "last": Object {
          "onClick": [Function],
          "value": 1,
        },
        "total": Object {
          "onClick": [Function],
          "value": 8,
        },
      }
    }
    focusedOn={
      Object {
        "normal": true,
      }
    }
    onTitleClick={[Function]}
    success={
      Object {
        "last": Object {
          "onClick": [Function],
          "value": 3,
        },
        "total": Object {
          "onClick": [Function],
          "value": 25,
        },
      }
    }
    timePeriod="24h"
    warning={
      Object {
        "last": Object {
          "onClick": [Function],
          "value": 2,
        },
        "total": Object {
          "onClick": [Function],
          "value": 20,
        },
      }
    }
  />
  <ScheduledTasksCard
    className=""
    focusedOn={
      Object {
        "normal": true,
      }
    }
    onClick={[Function]}
    scheduled={1}
    timePeriod="24h"
  />
</div>
`;
